src/ol/interaction/modifyinteraction.js

```diff
@@ -959,6 +1044,12 @@ ol.interaction.Modify.prototype.insertVertex_ = function(segmentData, vertex) {
       coordinates = geometry.getCoordinates();
       coordinates[depth[0]].splice(index + 1, 0, vertex);
       break;
+    case ol.geom.GeometryType.RECTANGLE:
```

Comment:
I think it's an overkill to be adding vertices to a rectangle which can only have 4.  It makes the code rather obscure, and I think it's the only reason you end up having to add another `origSegment` field to `ol.ModifySegmentDataType`, unless I'm missing something, which is easily possible.

You also don't have any comments about this, which makes it even harder to follow.

If you want, you can make the code for dragging snap points which are not part of the segment being modified more re-usable and share it for both the circle and rectangle features.